systemd-journal-upload: added custom http headers and compression

AndrewChubatiuk opened the pull request:
fixes https://github.com/systemd/systemd/issues/34568

poettering commented:
please drop the clang reformat stuff. it's not supposed to be hard enforced, it's a hint for new code only.

please prepare minimal commits, this is not reviewable otherwise.

poettering commented on src/journal-remote/journal-upload-journal.c:
```diff
@@ -303,6 +301,16 @@ static size_t journal_input_callback(void *buf, size_t size, size_t nmemb, void
+        int r = journal_input_data(buf, size * nmemb, 0, userp);
+        if (r < 0) {
+                return CURL_READFUNC_ABORT;
+        }
```
coding style, no {} around single line if blocks

poettering replied:
https://systemd.io/CODING_STYLE

poettering commented on src/journal-remote/journal-upload-journal.c:
```diff
@@ -303,6 +301,16 @@ static size_t journal_input_callback(void *buf, size_t size, size_t nmemb, void
+static size_t journal_input_callback(void *buf, size_t size, size_t nmemb, void *userp) {
+        assert(nmemb < SSIZE_MAX / size);
+        assert(!size_multiply_overflow(size, nmemb));
+        int r = journal_input_data(buf, size * nmemb, 0, userp);
```
"int r" always gets special treatment in our codebase, because we universally use such a local variable to store function error return codes in. Hence, we always declare it separately at the top of the function.

or in other words, do `int r;` at the top of the function

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -234,12 +263,44 @@ int start_upload(Uploader *u,
+                                readSize += input_fn(buf, step, size - step, data);
+                        }
+                        size_t csize;
+                        void *compressed = malloc(readSize);
```
our codebase must be able to handle OOM everywhere

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -681,6 +756,21 @@ static int parse_argv(int argc, char *argv[]) {
+                        if (r < 0)
+                                return log_error_errno(
+                                                r,
+                                                "--headers value should contain '='-delimited pairs, separated by comma");
```
in our docs and log messages we generally suffix cmdline params that take arguments with `=` to indicate they take a parameter. Hence, in this log msg write `--headers=` rather than `--headers`

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -681,6 +756,21 @@ static int parse_argv(int argc, char *argv[]) {
+                        if (!compression_supported(COMPRESSION_ZSTD))
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "ZSTD compression is not supported");
+                        arg_compress = true;
```
so, even if only zstd is supported initially: the parameter should be generic, i.e. probably be specified as --compress=zstd, to keep the door open for adding more later on

poettering commented on src/basic/headers-util.c:
```diff
@@ -0,0 +1,235 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
```
i don't thing headers-util.[ch] should be in basic/. AFAICS it's only useful within the journal-remote context, and hence should be moved there

poettering commented on src/core/load-fragment.c:
```diff
@@ -2905,6 +2906,79 @@ int config_parse_unset_environ(
         return 0;
 }
 
+int config_parse_headers(
```
src/core/ is the place for code touching pid1 itself. This is not be placed there, because it#s specific to journal-remote stuff.

poettering commented on src/journal-remote/journal-upload-journal.c:
```diff
@@ -284,14 +282,14 @@ static size_t journal_input_callback(void *buf, size_t size, size_t nmemb, void
                 if (filled == 0) {
                         log_error("Buffer space is too small to write entry.");
-                        return CURL_READFUNC_ABORT;
+                        return -1;
```
in our codebase it's not OK to make up literal error codes such as "-1". Always use propagate proper, negative errno-style error codes

poettering commented on src/journal-remote/journal-upload-journal.c:
```diff
@@ -249,25 +249,23 @@ static void check_update_watchdog(Uploader *u) {
                         if (r < 0) {
                                 log_error_errno(r, "Failed to move to next entry in journal: %m");
-                                return CURL_READFUNC_ABORT;
+                                return r;
```
you can shorten this to `return log_error_errno(r, …`, since log_error_errno() and related calls always return `-abs()` of the first arg you pass.

poettering commented on src/basic/headers-util.c:
```diff
@@ -0,0 +1,235 @@
+                } else {
+                        *d++ = *s;
+                }
+        } while (*(++s));
```
this will fall apart if you have something such as "Foo=bar=quux"

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -203,9 +213,28 @@ int start_upload(Uploader *u,
+                }
+
+                if (arg_headers != NULL) {
+                        for (size_t i = 0; arg_headers[i] != NULL; i++) {
```
STRV_FOREACH

poettering commented on src/basic/headers-util.c:
```diff
@@ -0,0 +1,235 @@
+char *convert_header(char *src) {
+        assert(src);
+        int len = strlen_ptr(src) + 2;
+        char *dest = malloc(len);
```
oom check

poettering commented on src/basic/headers-util.c:
```diff
@@ -0,0 +1,235 @@
+
+char *convert_header(char *src) {
+        assert(src);
+        int len = strlen_ptr(src) + 2;
```
strlen_ptr() is for cases where we want to gracefully handle NULL pointers. You already did an assert() on the pointer in the immediately preceeding line, hence strlen_ptr() is kinda pointless, you *know* it's not NULL here already, hence use plain strlen()

AndrewChubatiuk replied:
used strlen_ptr as it's already included here

poettering replied:
that's not really a reason..

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -234,12 +263,44 @@ int start_upload(Uploader *u,
+                                        size = step + readSize;
+                                }
+                                if (buf == NULL) {
+                                        buf = malloc(size * sizeof(char));
```
oom check

poettering replied:
realloc() is equivalent to malloc() btw, if first arg is NULL, hence the special case here is redundant

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -234,12 +263,44 @@ int start_upload(Uploader *u,
+                                } else {
+                                        void *newBuf = realloc(buf, size * sizeof(char));
+                                        if (newBuf == NULL) {
+                                                log_error_errno(SYNTHETIC_ERRNO(EXFULL),
```
the error code for OOM is ENOMEM.

Also use log_oom()

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -234,12 +263,44 @@ int start_upload(Uploader *u,
+                                        if (newBuf == NULL) {
+                                                log_error_errno(SYNTHETIC_ERRNO(EXFULL),
+                                                                "Failed to realloc buffer");
+                                                return -1;
```
don't make up errors, just do `return log_oom();` (as log_oom() returns -ENOMEM)

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -234,12 +263,44 @@ int start_upload(Uploader *u,
+                                        buf = malloc(size * sizeof(char));
+                                } else {
+                                        void *newBuf = realloc(buf, size * sizeof(char));
+                                        if (newBuf == NULL) {
```
coding style: for pointers, boolean, boolean-like fields we always do our if checks without explicit value comparison, i.e. use `if (!newBuf) …` here

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -234,12 +263,44 @@ int start_upload(Uploader *u,
+                                        buf = newBuf;
+                                }
+                                readSize += input_fn(buf, step, size - step, data);
+                        }
```
so if i am getting this right you are reading the whole journal into memory here? sorry, but that's not really acceptable, tis might be ridiculously large. compression needs to happen progressively: i.e. compress, write out, compress more, write more, compress even more, write even more

poettering commented on src/basic/headers-util.c:
```diff
@@ -0,0 +1,235 @@
+        return true;
+}
+
+char *convert_header(char *src) {
```
also, parameter should be const

poettering commented on src/basic/headers-util.c:
```diff
@@ -0,0 +1,235 @@
+        do {
+                if (*s == '=') {
+                        *d++ = ':';
+                        *d++ = ' ';
```
this is going to be  memory corruption if the input string contains multiple `=`

AndrewChubatiuk commented:
hey @poettering 
addressed your comments:
 - added configurable compression algorithm and level
 - making compression inside CURLOPT_READFUNCTION
 - remove headers pairs transformation
 - updated codestyle and added oom handling

poettering commented:
please marke the review items you addressed as "resolved"

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -9,8 +9,13 @@
 #include "sd-daemon.h"
 
+#include "math.h"
+#include "unit-printf.h"
```
Please drop unit-printf.h and load-fragment.h, and sort headers.
Also, math.h is a system header. Please include it earlier with `<>`.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +133,108 @@ static int check_cursor_updating(Uploader *u) {
+                void *data,
+                void *userdata) {
+
+        Compression* compression = ASSERT_PTR(data);
```
We use `Compression *compression`, rather than `Compression* compression`. That is, space before `*`.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +133,108 @@ static int check_cursor_updating(Uploader *u) {
+        assert(filename);
+        assert(rvalue);
+
+        if (isempty(rvalue)) {
+                *compression = COMPRESSION_NONE;
+        } else {
```
Return earlier:
```c
if (isempty(rvalue)) {
        *compression = COMPRESSION_NONE;
        return 0;
}
```

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +133,108 @@ static int check_cursor_updating(Uploader *u) {
+        if (isempty(rvalue)) {
+                *compression = COMPRESSION_NONE;
+        } else {
+                *compression = compression_from_string(ascii_strupper(strdup(rvalue)));
```
1. duplicated string is leaked,
2. missing error check in compression_from_string(),
3. do not assign unsupported type.
```c
_cleanup_free_ char *s = strdup(rvalue);
if (!s)
        return log_oom();

Compression c = compression_from_string(ascii_strupper(s));
if (c < 0)
        return log_syntax_parse_error(...);
if (!compression_supported(c))
        return log_syntax(unit, ...);

*compression = c;
return 0;
```

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +133,108 @@ static int check_cursor_updating(Uploader *u) {
+                void *data,
+                void *userdata) {
+
+        const Unit *u = userdata;
```
Please drop. `Unit` is for PID1, not used in journal-upload.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +133,108 @@ static int check_cursor_updating(Uploader *u) {
+                return 0;
+        }
+
+        /* If 'u' is set, we operate on the regular unit specifier table. Otherwise we use a manager-specific
+         * specifier table (in which case ltype must contain the runtime scope). */
+        const Specifier *table = u ? NULL : (const Specifier[]) {
+                COMMON_SYSTEM_SPECIFIERS,
+                COMMON_TMP_SPECIFIERS,
+                COMMON_CREDS_SPECIFIERS(ltype),
+                { 'h', specifier_user_home,  NULL },
+                { 's', specifier_user_shell, NULL },
+                {}
+        };
```
I do not think this is necessary. Please drop.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +133,108 @@ static int check_cursor_updating(Uploader *u) {
+                if (r == 0)
+                        return 0;
+
+                if (table)
+                        r = specifier_printf(word, sc_arg_max(), table, NULL, NULL, &resolved);
+                else
+                        r = unit_env_printf(u, word, &resolved);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
+                                   "Failed to resolve specifiers in %s, ignoring: %m", word);
+                        continue;
+                }
```
I do not think it is necessary to support specifier. What's the use case? Please drop this block.

yuwata commented on src/journal-remote/headers-util.c:
```diff
@@ -0,0 +1,207 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
```
This file is mostly a subset of env-util.c .
Please try to use env-util.c .

AndrewChubatiuk replied:
it's not a subset, it has differences

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +133,108 @@ static int check_cursor_updating(Uploader *u) {
+        for (const char *p = rvalue;; ) {
+                _cleanup_free_ char *word = NULL, *resolved = NULL;
+
+                r = extract_first_word(&p, &word, NULL, EXTRACT_CUNESCAPE|EXTRACT_UNQUOTE);
```
I do not think it is necessary to support multiple header entries per setting.
Please drop the loop and `extract_first_word().`

AndrewChubatiuk replied:
left extract_first_word for unquoting and unescaping

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -279,13 +432,16 @@ int start_upload(Uploader *u,
                 return log_error_errno(SYNTHETIC_ERRNO(EXFULL),
                                        "curl_easy_setopt CURLOPT_URL failed: %s",
                                        curl_easy_strerror(code));
-
```
Please unnecessarily touch the code. Please keep the empty line.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -349,6 +505,7 @@ static int dispatch_fd_input(sd_event_source *event,
                 return 0;
         }
 
+
```
doubled empty line.

yuwata reviewed (changes requested):
Not fully reviewed, but several random comments.

Please split the huge commit into small ones. At least, 
1. first introduce the level support for compress.c,
2. then change journal-upload.c .

Please update man page, add test cases.

AndrewChubatiuk commented:
hey @yuwata
I've updated a PR, made two commits and addressed your comments

yuwata commented on src/journal-remote/meson.build:
```diff
@@ -88,6 +99,16 @@ executables += [
+executables += [
+        test_template + {
+                'sources' : files('test-header-util.c'),
+                'link_with' : [
+                        libshared,
+                        libsystemd_journal_upload,
+                ],
```
Not necessary. Please drop. And, add header-util.c to `files()`:
```
'sources' : files('test-header-util.c',
                  'header-util.c')
```

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -423,11 +551,40 @@ static int setup_uploader(Uploader *u, const char *url, const char *state_file)
+        sev.sigev_notify_attributes = 0;
+
+        if (arg_batch_timeout_usec != USEC_INFINITY) {
+                r = timer_create(CLOCK_REALTIME, &sev, &u->batch_timeout_id);
```
BTW, please use `sd_event_add_time()` or friend.

yuwata reviewed (changes requested):
The second commit is still too large. Please split into small pieces.
Also, please read our coding style documentation https://systemd.io/CODING_STYLE/ .

AndrewChubatiuk commented on src/journal-remote/journal-upload.h:
```diff
@@ -3,12 +3,18 @@
 #pragma once
 
 #include <inttypes.h>
+#include <stdatomic.h>
```
using atomic bool for uploading variable, as read/write can happen simultaneously to it in multiple places at the same time

poettering replied:
uhh? no, we do not use threads.

AndrewChubatiuk commented on src/journal-remote/journal-upload-journal.c:
```diff
@@ -284,22 +293,35 @@ static size_t journal_input_callback(void *buf, size_t size, size_t nmemb, void
                         return CURL_READFUNC_ABORT;
                 filled += w;
 
+                update_uploader_stat(u, w);
```
batch stat modification is happening in this method, if events or bytes limit is reached (which can be configured using BatchMaxSize and BatchMaxEvents), it sets `u->uploading` to false and forces curl to send data

poettering commented on man/journal-upload.conf.xml:
```diff
@@ -60,6 +60,54 @@
+
+        <listitem><para>Max payload size of journal entries batch in a single HTTP request.</para>
+
+        <xi:include href="version-info.xml" xpointer="v232"/></listitem>
```
the version numbers are not right. it's v257 (or maybe v258, let's see)

poettering commented on man/journal-upload.conf.xml:
```diff
@@ -60,6 +60,54 @@
+      <varlistentry>
+        <term><varname>Header=</varname></term>
+
+        <listitem><para>HTTP header which should be added to each request to URL.</para>
```
please be more precise, i.e. say in which format, maybe provide an example

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,166 @@
+        if (!l)
+                return NULL;
+
+        /* Drops every occurrence of the heaer var setting p in the
```
heaer → header

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,166 @@
+                return -EINVAL;
+        }
+
+        name = strndupa_safe(p, t - p);
```
please don't use strndupa_safe() (i.e. stack memory) for something user-controlled like this. We try to get away from that

AndrewChubatiuk replied:
please advise here

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,166 @@
+                }
+
+        /* We didn't find a match, we need to append p or create a new strv */
+        r = strv_extend(l, p);
```
consume

poettering commented on src/journal-remote/journal-upload-journal.c:
```diff
@@ -251,9 +251,18 @@ static void check_update_watchdog(Uploader *u) {
+        if (u->compression != COMPRESSION_NONE) {
+                src = malloc(sizeof(char) * size * nmemb);
+                if (!src) {
+                        log_error("Failed to initialize compression buffer.");
```
log_oom()

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -156,6 +234,26 @@ static int update_cursor_state(Uploader *u) {
+                its.it_value.tv_sec = arg_batch_timeout_usec / USEC_PER_SEC;
+                its.it_value.tv_nsec = (arg_batch_timeout_usec % USEC_PER_SEC) * 1000;
+                its.it_interval.tv_sec = 0;
+                its.it_interval.tv_nsec = 0;
```
please use structured initialization, i.e.: 

```c
struct itimerspec its = {};
```

And then use timespec_load() to convert from usec_t to timespec

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -203,6 +303,23 @@ int start_upload(Uploader *u,
+                if (arg_compression != COMPRESSION_NONE) {
+                        char *header;
+                        char *encoding = strdup(compression_to_string(arg_compression));
+                        asprintf(&header, "Content-Encoding: %s", ascii_strlower(encoding));
```
asprintf() my fay fail.

also, please use strjoin() to join strings

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -423,11 +551,40 @@ static int setup_uploader(Uploader *u, const char *url, const char *state_file)
         if (r < 0)
                 return log_error_errno(r, "Failed to install SIGINT/SIGTERM handlers: %m");
 
+        struct sigevent sev;
```
use structured initialization, i.e.

```c
struct sigevent sev = {
        .sigev_notify = SIGEV_THREAD,
        …
}
```

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -423,11 +551,40 @@ static int setup_uploader(Uploader *u, const char *url, const char *state_file)
+                r = timer_create(CLOCK_REALTIME, &sev, &u->batch_timeout_id);
+                if (r < 0) {
+                        return log_error_errno(errno, "Failed to create a batch timer: %m");
+                }
```
no {} around single line if blocks, as per coding style

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -423,11 +551,40 @@ static int setup_uploader(Uploader *u, const char *url, const char *state_file)
         return load_cursor_state(u);
 }
 
+void update_uploader_stat(Uploader *u, int w) {
```
"int" is a really weird type here. should by uint64_t i guess

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -423,11 +551,40 @@ static int setup_uploader(Uploader *u, const char *url, const char *state_file)
+        sev.sigev_notify = SIGEV_THREAD;
+        sev.sigev_value.sival_ptr = u;
+        sev.sigev_notify_function = &finish_uploading;
+        sev.sigev_notify_attributes = 0;
```
sorry, what is this? nope. we do not use threads for this, and unix signals are the worst. i don#t get this. this has to go.

poettering commented on src/journal-remote/journal-upload.h:
```diff
@@ -49,10 +55,16 @@ typedef struct Uploader {
         const char *state_file;
 
+        double bytes_left;
+        double entries_left;
```
sorry, but no way. we are not going to count bytes in floating point. use uint64_t or so.

poettering commented on src/journal-remote/test-header-util.c:
```diff
@@ -0,0 +1,41 @@
+        assert_se(strv_length(headers) == 3);
+
+        in = strdup("Name: Rewrite");
+        out = strdup(in);
```
check for all oom cases. in test code via asserts in some form

yuwata reviewed (changes requested):
Please split the second commit into at least three: for batch, compression, and header.

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -58,8 +62,14 @@ static bool arg_merge = false;
+static Compression arg_compression = COMPRESSION_NONE;
+static int arg_compression_level = -1;
+static uint64_t arg_batch_max_bytes = UINT_MAX;
+static uint64_t arg_batch_max_entries = UINT_MAX;
```
the max for uint64_t is UINT64_MAX, not UINT_MAX

poettering replied:
only half addressed

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +131,83 @@ static int check_cursor_updating(Uploader *u) {
+
+        Compression c = compression_from_string(ascii_strupper(s));
+        if (c < 0)
+                log_syntax_parse_error(unit, filename, line, c, lvalue, rvalue);
```
you should leave the function here

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +131,83 @@ static int check_cursor_updating(Uploader *u) {
+        r = strv_header_replace_consume(headers, TAKE_PTR(unescaped));
+        if (r < 0)
+                return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                  "Failed to updated headers: %s", rvalue);
```
updated → updae

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -203,6 +311,27 @@ int start_upload(Uploader *u,
                 h = l;
 
+                if (arg_compression != COMPRESSION_NONE) {
+                        char *header;
```
`header` is currently leaked. should have `_cleanup_free_` too, and NULL initialization

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -279,6 +408,10 @@ int start_upload(Uploader *u,
 
+        r = refresh_timeout(u);
+        if (r < 0)
+                return log_error_errno(r, "Failed to refresh uploader timer");
```
duplicate logging. refresh_timeout() already logs about all errors afaics

AndrewChubatiuk replied:
replaced it with a plain code return

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -302,10 +438,37 @@ static size_t fd_input_callback(void *buf, size_t size, size_t nmemb, void *user
+                update_uploader_stat(u, n);
+                if (u->compression == COMPRESSION_NONE) {
+                        return n;
+                }
```
no {} around single line if blocks, as per coding style

https://systemd.io/CODING_STYLE/#formatting

AndrewChubatiuk replied:
sorry, always forgetting about this

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -302,10 +438,37 @@ static size_t fd_input_callback(void *buf, size_t size, size_t nmemb, void *user
+
+                r = compress_blob(u->compression, compression_buffer, n, buf, read_size, &compressed_size, u->compression_level);
+                if (r < 0) {
+                        log_error_errno(r, "Failed to compress %ld bytes using Compression=%s, CompressionLevel=%d)",
```
always suffix log messages that are about some error code you receive with `: %m`

AndrewChubatiuk replied:
updated

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -580,6 +786,8 @@ static int parse_argv(int argc, char *argv[]) {
                 { "follow",       optional_argument, NULL, ARG_FOLLOW         },
                 { "save-state",   optional_argument, NULL, ARG_SAVE_STATE     },
+                { "compress",     optional_argument, NULL, ARG_COMPRESSION    },
+                { "header",       optional_argument, NULL, ARG_HEADER         },
```
these two switches need a man page update too

AndrewChubatiuk replied:
removed these cmd flags as they are not that useful

AndrewChubatiuk commented on src/journal-remote/journal-upload.c:
```diff
@@ -681,6 +889,21 @@ static int parse_argv(int argc, char *argv[]) {
+                        break;
+
+                case ARG_COMPRESSION:
+                        arg_compression = compression_from_string(ascii_strupper(optarg));
```
what do you suggest as an alternative?

poettering replied:
insist on the lowercase form, i.e. just drop the ascii_strupper() without replacement.

Our user-facing enums generally are case sensitive, not sure why we should depart from that here.

AndrewChubatiuk replied:
`compression_from_string` requires upper case string

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +131,83 @@ static int check_cursor_updating(Uploader *u) {
+                return 0;
+        }
+
+        if (!header_is_valid(rvalue)) {
```
if this is an escaped string you should unescape first, and then validate it as a header

poettering commented on src/journal-remote/journal-upload.c:
```diff
@@ -203,6 +311,27 @@ int start_upload(Uploader *u,
                 h = l;
 
+                if (arg_compression != COMPRESSION_NONE) {
+                        _cleanup_free_ char *header = NULL;
```
move down, where you need it?

poettering commented:
i don't grok the "batch" knobs. Why do we need those?

AndrewChubatiuk commented:
> i don't grok the "batch" knobs. Why do we need those?

they are not useful in case of small logs stream, but for huge one these settings should allow to cut data in batches depending on configuration

poettering commented:
> > i don't grok the "batch" knobs. Why do we need those?
> 
> they are not useful in case of small logs stream, but for huge one these settings should allow to cut data in batches depending on configuration

batches of what?

i mean, TCP deals with this kinda of stuff on its own? why do we have to care? what is the "Batch" stuff precisely doing?

AndrewChubatiuk commented:
> batches of what?

batches of journal entries

yuwata commented on src/basic/compress.c:
```diff
@@ -237,7 +245,7 @@ int compress_blob_lz4(const void *src, uint64_t src_size,
                 return -ENOBUFS;
 
-        r = sym_LZ4_compress_default(src, (char*)dst + 8, src_size, (int) dst_alloc_size - 8);
+        r = sym_LZ4_compress_HC(src, (char*)dst + 8, src_size, (int) dst_alloc_size - 8, level);
```
Hm, is this change algorithm, especially picks slower one??

The documentation says level must be positive, so, please check the level.

yuwata replied:
Yes, but should we pick any slower algorithm? I do not think so, at least for journal.

AndrewChubatiuk replied:
what can you suggest?

yuwata replied:
How about the following? Use `_default()` when level is neative or 0, otherwise use the slower algorithm with the specified level.

yuwata commented on src/test/test-compress.c:
```diff
@@ -242,7 +242,7 @@ static void test_lz4_decompress_partial(void) {
         huge[HUGE_SIZE - 1] = '\0';
 
-        r = sym_LZ4_compress_default(huge, buf, HUGE_SIZE, buf_size);
+        r = sym_LZ4_compress_HC(huge, buf, HUGE_SIZE, buf_size, /* level = */ 0);
```
Please use `-1` for consistency with others.

AndrewChubatiuk replied:
this is not `compress_blob` wrapper, so suppose it's better to set 0 here

yuwata replied:
Ah, indeed, then use 1, the level needs to be positive.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +124,39 @@ static int check_cursor_updating(Uploader *u) {
+                void *data,
+                void *userdata) {
+
+        char *compression = ASSERT_PTR(data);
+        if (isempty(rvalue)) {
```
Please insert an empty line between variable declaration(s) and other code.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +124,39 @@ static int check_cursor_updating(Uploader *u) {
+                return 0;
+        }
+
+        _cleanup_free_ char *s = strdup(rvalue);
+        if (!s)
+                return log_oom();
```
Not necessary anymore. Please drop this, and directly pass `rvalue` to `compression_from_string()`.

Also, currently all entries in `compression_table` are in upper, but please make it lower. Please do that in a different commit. Then, update docs/ENVIRONMENT.md .

AndrewChubatiuk replied:
do you mean add to a table lower case values of replace existing ones with lower case?

yuwata replied:
I think it is ok to make the existing table lower.

bluca replied:
This breaks backward compatibility - it's ok to support both, but the parser needs to be case-insensitive then

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +124,39 @@ static int check_cursor_updating(Uploader *u) {
+                return log_syntax_parse_error(unit, filename, line, c, lvalue, rvalue);
+        if (!compression_supported(c))
+                return log_syntax(unit, LOG_WARNING, filename, line, 0,
+                           "Compression=%s is not supported on a system", rvalue);
```
Spurious indentation. Please align with `(`.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -121,6 +124,39 @@ static int check_cursor_updating(Uploader *u) {
         return 0;
 }
 
+int config_parse_compression(
```
This is quite generic. Please move this to src/shared/conf-parser.c .

yuwata replied:
I guess, it was not generic at that time.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -203,6 +239,22 @@ int start_upload(Uploader *u,
+                                return log_oom();
+
+                        _cleanup_free_ char *header = NULL;
+                        header = strjoin("Content-Encoding: ", ascii_strlower(encoding));
```
By making the `compression_table` lower, `ascii_strlower()` and strdup() can be dropped.

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -60,6 +60,70 @@
         <xi:include href="version-info.xml" xpointer="v232"/></listitem>
```
Please move the man page update to the relevant commits.

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -60,6 +60,70 @@
+      <varlistentry>
+        <term><varname>Compression=</varname></term>
+
+        <listitem><para>Compression algorithm to apply to logs data before sending.
```
`to apply` -> `to be applied` ? Not sure, I am not good at English.

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -60,6 +60,70 @@
+
+        <listitem><para>Compression algorithm to apply to logs data before sending.
+        Supported values are <literal>ZSTD</literal>, <literal>XZ</literal>, <literal>LZ4</literal> and <literal>NONE</literal>.
+        Default is <literal>NONE</literal></para>
```
- Please make them lower case, when you update `compression_table`.
- Please insert a comma before `and`.
- `Default is` -> `Defaults to`
- Missing period at the and.

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -60,6 +60,70 @@
+      <varlistentry>
+        <term><varname>CompressionLevel=</varname></term>
+
+        <listitem><para>Compression level is an integer, which defines a level of journal entries compression.
```
Maybe, `Specifies the compression level of data to be uploaded.` is better ?

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -60,6 +60,70 @@
+
+        <listitem><para>Compression level is an integer, which defines a level of journal entries compression.
+        Each algorithm has it's own range of allowed values:  XZ - [1 - 9], ZSTD - [1 - 22], LZ4 - [1 - 9].
+        Default value: -1, which set's a default compression level for each compresion algorithm.</para>
```
`Defaults to -1, and the default compression level for each compression algorithm will be used.`

Please also mention that this will be ignored when `CompressionLevel=none`.

yuwata reviewed (changes requested):
Reviewed only compression related commits.
I think it is better to move the header and batch things to another later PR(s).

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -60,6 +60,26 @@
+        <term><varname>CompressionLevel=</varname></term>
+
+        <listitem><para>Specifies the compression level of data to be uploaded.
+        Each algorithm has it's own range of allowed integer values:  xz - [1 - 9], zstd - [1 - 22], lz4 - [1 - 9].
+        Defaults to -1, and the default compression level for each compression algorithm will be used.</para>
```
Maybe, better to not mention the detailed range, but just say this takes an integer and acceptable range depends on the algorithm. Also, please drop the internal default `-1`, but say `Defaults to unset, ...`.

Please mention that this setting will be ignored when `Compression=none`.

yuwata commented on src/test/test-compress-benchmark.c:
```diff
@@ -160,13 +160,13 @@ int main(int argc, char *argv[]) {
-                test_compress_decompress("XZ", i, compress_blob_xz, decompress_blob_xz);
+                test_compress_decompress("xz", i, compress_blob_xz, decompress_blob_xz);
 #endif
 #if HAVE_LZ4
-                test_compress_decompress("LZ4", i, compress_blob_lz4, decompress_blob_lz4);
+                test_compress_decompress("lz4", i, compress_blob_lz4, decompress_blob_lz4);
 #endif
 #if HAVE_ZSTD
-                test_compress_decompress("ZSTD", i, compress_blob_zstd, decompress_blob_zstd);
+                test_compress_decompress("zstd", i, compress_blob_zstd, decompress_blob_zstd);
```
These changes are not necessary, the strings are used only in the log message. Please drop.

yuwata commented on src/test/test-compress.c:
```diff
@@ -292,79 +292,79 @@ int main(int argc, char *argv[]) {
         random_bytes(data + 7, sizeof(data) - 7);
 
 #if HAVE_XZ
-        test_compress_decompress("XZ", compress_blob_xz, decompress_blob_xz,
```
Also here, and below.

yuwata commented on src/basic/compress.c:
```diff
@@ -109,10 +111,10 @@ static inline void lzma_end_wrapper(lzma_stream *ls) {
-        [COMPRESSION_NONE] = "NONE",
-        [COMPRESSION_XZ]   = "XZ",
-        [COMPRESSION_LZ4]  = "LZ4",
-        [COMPRESSION_ZSTD] = "ZSTD",
+        [COMPRESSION_NONE] = "none",
+        [COMPRESSION_XZ]   = "xz",
+        [COMPRESSION_LZ4]  = "lz4",
+        [COMPRESSION_ZSTD] = "zstd",
 };
```
Could you split out this change (and update for docs/ENVIRONMENT.md) to another commit? This is irrelevant to the compression level support.

yuwata reviewed (changes requested):
Could you drop the header and batch things from this PR?
Please open another PR after the compression support is merged.

yuwata commented:
The compression part is moved to #34822, making this PR draft, until #34822 is merged.

yuwata commented:
#36334 is merged. Please rebase.

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,170 @@
+                       "|"       \
+                       "~"       \
+                       "^"       \
+                       "%"
```
pleas reduce the line breaks a bit here. just write this out is a simple string

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,170 @@
+                       "%"
+
+static bool header_name_is_valid(const char *e, size_t n) {
+        if (!e)
```
this check should be n == 0 instead,

and please add `assert(e)` afterwards.

because if n is > 0 and e is NULL that's a programming error, and we should assert() on those

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,170 @@
+        if (!e)
+                return false;
+
+        int n = strlen_ptr(e);
```
simple strlen_ptr() would suffice too

poettering replied:
you drop 32bit on the floor here. strlen() returns size_t (i.e. typically 64bit), not `int` (typically 32bit). change type to `size_t`

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,170 @@
+
+        for (const char *p = e; p < e + n; p++)
+                if (!strchr(VALID_HEADER_VALUE_CHARS, *p))
+                        return false;
```
in_charset()

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,170 @@
+        int n = strlen_ptr(e);
+
+        if (n < 0)
+                return false;
```
this is impossible

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,170 @@
+
+bool header_is_valid(const char *e) {
+        const char *eq;
+
```
maybe do:

```c
if (isempty(e))
        return false;
```

poettering commented on man/journal-upload.conf.xml:
```diff
@@ -129,6 +129,25 @@
+        <term><varname>Header=</varname></term>
+
+        <listitem><para>HTTP header which should be added to each request to URL. Headers are expected to
+        be unquoted and it's expected to be no more than one header per each option.</para>
```
hmm Set-Cookie is often used multiple times within the same request. any particular reason why to be stricter here? why not just allow that?

poettering replied:
"unquoted", what do you mean by that?

you should mention that C-style escapes are supported.

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,170 @@
+        return l;
+}
+
+int strv_header_replace_consume(char ***l, char *p) {
```
i am really not convinced we should erase earlier entries with the same header field name. i think we should just append to the end, accepting duplicates

AndrewChubatiuk replied:
you are right, updated implementation

poettering commented on src/journal-remote/header-util.c:
```diff
@@ -0,0 +1,170 @@
+
+        name = strndupa(p, t - p);
+        if (!name) {
+                free(p);
```
please do not use manual `free(p)` multiple times. instead use `_cleanup_free_`, and TAKE_PTR in order to defuse the destruction

poettering commented on man/journal-upload.conf.xml:
```diff
@@ -129,6 +129,29 @@
+          Header=HeaderName: HeaderValue
+          Header=HeaderName: NewValue
+          Header=HeaderName: HeaderValue
+        </programlisting>
```
hmm, some renderings of docbook really care about leading whitespace, hence please drop any leading whitespace within `<programlisting>` tags. and ideally even place opening tag on same line as first contents line without separating whitespace, and the closing line after it without separating whitespace

poettering commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,96 @@
+                return log_oom();
+
+        return r;
+}
```
this function is so low-level. i think it's better to do error logging in caller, not in here

poettering commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,96 @@
+                return 0;
+
+        if (!header_name_is_valid(name))
+                return 0;
```
these should not be eaten up silently. please catch these in config_parse_header() and do proper log_syntaxt() loggging about them!

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,92 @@
+- Alphanumeric characters: a-z, A-Z, and 0-9
+- The following special characters: - and _ */
+#define VALID_HEADER_NAME_CHARS \
+        DIGITS LETTERS "_-"
```
`DIGITS LETTERS` -> `ALPHANUMERICAL`

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,92 @@
+#define VALID_HEADER_NAME_LENGTH 40
+
+#define VALID_HEADER_VALUE_CHARS \
+        DIGITS LETTERS "_ :;.,\\/'\"?!(){}[]@<>=-+*#$&`|~^%"
```
Also here

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,92 @@
+        if (!e)
+                return false;
+
+        if (strlen(e) > VALID_HEADER_NAME_LENGTH)
```
zero length is OK??

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,92 @@
+
+        *t++ = '\0';
+
+        r = header_put(headers, TAKE_PTR(unescaped), skip_leading_chars(t, WHITESPACE));
```
Please drop `TAKE_PTR()`. That causes memory leak.

yuwata replied:
Maybe, better to chop the leading and trailing spaces in header name, to support something like the following:
```
Header=  NAME  :  VALUE
```

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,92 @@
+
+        r = cunescape(rvalue, 0, &unescaped);
+        if (r < 0)
+                return log_syntax(unit, LOG_WARNING, filename, line, r,
+                                  "Failed to unescape headers: %s", rvalue);
```
Please do not propagate error in such case:
```c
if (r < 0) {
        log_syntax(...);
        return 0;
}
```

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,92 @@
+
+        t = strchr(unescaped, ':');
+        if (!t)
+                return log_syntax(unit, LOG_WARNING, filename, line, EINVAL,
```
Same here.

Also, please use `0` rather than `EINVAL`.

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,92 @@
+
+        r = header_put(headers, TAKE_PTR(unescaped), skip_leading_chars(t, WHITESPACE));
+        if (r < 0)
+                return log_syntax(unit, LOG_WARNING, filename, line, r,
```
Also here.

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -226,6 +230,18 @@ int start_upload(Uploader *u,
+                char **values;
+                const char *name;
+                ORDERED_HASHMAP_FOREACH_KEY(values, name, arg_headers) {
+                        _cleanup_free_ char *header = strjoin(name, ": ", strv_join(values, ", "));
```
memory leak and missing error check.
Please check return value of `strv_join()`, and free the string after use.
```c
_cleanup_free_ char *joined = strv_join(values, ", ");
if (!joined)
        return log_oom();

_cleanup_free_ char *header = strjoin(name, ": ", joined);
if (!header)
        return log_oom();
```

yuwata commented on src/journal-remote/test-journal-header-util.c:
```diff
@@ -0,0 +1,17 @@
+TEST(header_put) {
+        _cleanup_ordered_hashmap_free_ OrderedHashmap *headers = NULL;
+
+        ASSERT_EQ(header_put(&headers, "NewName", "Val"), 1);
```
Please use `ASSERT_OK_POSITIVE()`

yuwata commented on src/journal-remote/test-journal-header-util.c:
```diff
@@ -0,0 +1,17 @@
+        ASSERT_EQ(header_put(&headers, "NewName", "Val"), 1);
+        ASSERT_EQ(header_put(&headers, "Name", "FirstName"), 1);
+        ASSERT_EQ(header_put(&headers, "Name", "Override"), 1);
+        ASSERT_EQ(header_put(&headers, "Name", "FirstName"), 0);
```
Please use `ASSERT_OK_ZERO()` here.

yuwata commented on src/journal-remote/test-journal-header-util.c:
```diff
@@ -0,0 +1,17 @@
+        ASSERT_EQ(header_put(&headers, "Name", "FirstName"), 1);
+        ASSERT_EQ(header_put(&headers, "Name", "Override"), 1);
+        ASSERT_EQ(header_put(&headers, "Name", "FirstName"), 0);
+        ASSERT_ERROR(header_put(&headers, "InvalidN@me", "test"), EINVAL);
```
Please add more failure cases, passing an invalid value, `NULL`, and so on.

yuwata reviewed (changes requested):
Sorry for late review.

Please also add test cases in a new TEST-04-JOURNAL.journal-update.sh or so.

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -129,6 +129,28 @@
+        <para>Header name can contains alphanumeric values, "_" and "-" symbols additionally.</para>
+
+        <para>This option may be specified more than once, in which case all listed headers will be set. If
+        the same header name is listed more than once, all its unique values will be concatenated with comma.
```
Please mention that an empty string clears all previous assignments.
Also, please mention that the header name and value must be separated with ':'.

bluca commented:
```
[ 4363.711793] systemd-journal-upload[17611]: ==17611==ERROR: LeakSanitizer: detected memory leaks
[ 4363.711793] systemd-journal-upload[17611]: Direct leak of 22 byte(s) in 1 object(s) allocated from:
[ 4363.712724] systemd-journal-upload[17611]:     #0 0x7f7ad96d5a83 in malloc (/usr/lib/clang/19/lib/x86_64-redhat-linux-gnu/libclang_rt.asan.so+0xd5a83) (BuildId: 8830e46c5a6c721fccc9df83c6a81de1cafbccb0)
[ 4363.712724] systemd-journal-upload[17611]:     #1 0x7f7ad8b7a91e in cunescape_length_with_prefix (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x77a91e) (BuildId: 8101976f4029915bdd19abdd25ed4e0e6d0e4909)
[ 4363.712724] systemd-journal-upload[17611]:     #2 0x556086eafa15 in cunescape_length /usr/src/debug/systemd/src/basic/escape.h:53:16
[ 4363.712724] systemd-journal-upload[17611]:     #3 0x556086eafa15 in cunescape /usr/src/debug/systemd/src/basic/escape.h:56:16
[ 4363.712724] systemd-journal-upload[17611]:     #4 0x556086eafa15 in config_parse_header /usr/src/debug/systemd/src/journal-remote/journal-header-util.c:74:13
[ 4363.712724] systemd-journal-upload[17611]:     #5 0x7f7ad87f797b  (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x3f797b) (BuildId: 8101976f4029915bdd19abdd25ed4e0e6d0e4909)
[ 4363.712724] systemd-journal-upload[17611]:     #6 0x7f7ad87f66b3 in config_parse (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x3f66b3) (BuildId: 8101976f4029915bdd19abdd25ed4e0e6d0e4909)
[ 4363.712724] systemd-journal-upload[17611]:     #7 0x7f7ad87f9c2d in config_parse_many (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x3f9c2d) (BuildId: 8101976f4029915bdd19abdd25ed4e0e6d0e4909)
[ 4363.712724] systemd-journal-upload[17611]:     #8 0x7f7ad87fa841 in config_parse_standard_file_with_dropins_full (/usr/lib64/systemd/libsystemd-shared-258-devel.so+0x3fa841) (BuildId: 8101976f4029915bdd19abdd25ed4e0e6d0e4909)
[ 4363.712724] systemd-journal-upload[17611]:     #9 0x556086ea8a1b in config_parse_standard_file_with_dropins /usr/src/debug/systemd/src/shared/conf-parser.h:127:16
[ 4363.712724] systemd-journal-upload[17611]:     #10 0x556086ea8a1b in parse_config /usr/src/debug/systemd/src/journal-remote/journal-upload.c:687:16
[ 4363.712724] systemd-journal-upload[17611]:     #11 0x556086ea8a1b in run /usr/src/debug/systemd/src/journal-remote/journal-upload.c:938:13
[ 4363.712724] systemd-journal-upload[17611]:     #12 0x556086ea8a1b in main /usr/src/debug/systemd/src/journal-remote/journal-upload.c:1017:1
[ 4363.712724] systemd-journal-upload[17611]:     #13 0x7f7ad8210247 in __libc_start_call_main (/lib64/libc.so.6+0x3247) (BuildId: 9059c4f4ffc1a63e796d41f813ed48baa3ad4096)
[ 4363.712724] systemd-journal-upload[17611]:     #14 0x7f7ad821030a in __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x330a) (BuildId: 9059c4f4ffc1a63e796d41f813ed48baa3ad4096)
[ 4[363](https://github.com/systemd/systemd/actions/runs/13837091958/job/38714786300?pr=34595#step:13:364).712724] systemd-journal-upload[17611]:     #15 0x556086e9fc54  (/usr/lib/systemd/systemd-journal-upload+0xac54) (BuildId: e7edf3a802f2d113e8b0ef558f7069fd4a4501b5)
[ 4363.713824] systemd-journal-upload[17611]: SUMMARY: AddressSanitizer: 22 byte(s) leaked in 1 allocation(s).
```

bluca commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,98 @@
+                void *userdata) {
+
+        OrderedHashmap **headers = ASSERT_PTR(data);
+        _cleanup_free_ char *unescaped, *t;
```
unescaped needs to be initialized to NULL, and t cannot have _cleanup_ so it needs to be on a different declaration line

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -129,6 +129,31 @@
+
+        <para>This option may be specified more than once, in which case all listed headers will be set. If
+        the same header name is listed more than once, all its unique values will be concatenated with comma.
+        Setting <literal>Header</literal> to empty string clears all previous assignments.
```
Please use varname tag, and suffixed with `=`:
`<varname>Header=<varname>`

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,99 @@
+        ALPHANUMERICAL "_ :;.,\\/'\"?!(){}[]@<>=-+*#$&`|~^%"
+
+static bool header_name_is_valid(const char *e) {
+        if (!e)
+                return false;
+
+        if (strlen(e) == 0 || strlen(e) > VALID_HEADER_NAME_LENGTH)
```
```c
if (isempty(e))
        return false;

if (strlen(e) > VALID_HEADER_NAME_LENGTH)
        return false;
```

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,99 @@
+#define VALID_HEADER_NAME_CHARS \
+        ALPHANUMERICAL "_-"
+
+#define VALID_HEADER_NAME_LENGTH 40
```
Please rename this to HEADER_NAME_LENGTH_MAX

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,99 @@
+#define VALID_HEADER_NAME_LENGTH 40
+
+#define VALID_HEADER_VALUE_CHARS \
+        ALPHANUMERICAL "_ :;.,\\/'\"?!(){}[]@<>=-+*#$&`|~^%"
```
Where it is defined? Could you add a reference in the comment?

Also, are there no limit in the length?

AndrewChubatiuk replied:
i haven't found any limits in a standard, but per standard it's allowed to set limits and respond with http 400 in case of too long header names and values
standard allows US-ASCII characters for headers, can i use `is_ascii` there?
this list of symbols I've took from cloudflare website, thought they are relying on a RFC

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,99 @@
+        if (!e)
+                return false;
+
+        return in_charset(e, VALID_HEADER_VALUE_CHARS);
```
An empty value is allowed?

AndrewChubatiuk replied:
https://datatracker.ietf.org/doc/html/rfc7231#section-5.3.4
it's mentioned that Accept-Encoding header can be empty

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -226,6 +230,23 @@ int start_upload(Uploader *u,
+                        if (!joined)
+                                return log_oom();
+
+                        _cleanup_free_ char *header = strjoin(name, ": ", joined);
```
An empty value string is allowed? If allowed, and `joined` is an empty string, then still `": "` is necessary?

AndrewChubatiuk replied:
colon is still necessary

yuwata replied:
How about the trailing space in that case?

AndrewChubatiuk replied:
https://www.rfc-editor.org/rfc/rfc2616#section-4.2
```
The field value MAY be preceded by any amount of LWS, though a single SP is preferred
```

yuwata replied:
Please also check the validity of the concatenated string with `header_value_is_valid()`.

AndrewChubatiuk replied:
concatenated header contains pair and header_value_is_valid checks validates value only

yuwata replied:
I mean, `header_value_is_valid(joined)`.

yuwata reviewed (changes requested):
Several questions and minor requests.

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,109 @@
+
+#define HEADER_NAME_LENGTH_MAX 40
+
+#define HEADER_VALUE_LENGTH_MAX 8000
```
Please mention that no RFC seem to define that, and this is just for safety.

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,109 @@
+
+bool header_value_is_valid(const char *e) {
+        if (!e)
+                return false;
+        if (strlen(e) > HEADER_VALUE_LENGTH_MAX)
```
Please insert an empty line for consistency with other places.

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,109 @@
+        r = cunescape(rvalue, 0, &unescaped);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to unescape headers: %s", rvalue);
```
Please add `", ignoring"` in the log message like the below.

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,109 @@
+        r = header_put(headers, strstrip(unescaped), skip_leading_chars(t, WHITESPACE));
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to update headers: %s", rvalue);
```
Also here. Please add `", ignoring"`

yuwata commented on src/journal-remote/journal-upload.c:
```diff
@@ -226,6 +230,26 @@ int start_upload(Uploader *u,
+
+                        if (header_value_is_valid(joined))
+                                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                                       "Invalid header value %s for %s", joined, name);
```
The condition is swapped...

I do not think this should be critical.
Please log in warning level, and ignore the failure.
Also, `joined` is spurious, in most cases, it is too long. Hence, please do not include `joined` in the log message.
```c
if (!header_value_is_valid(joined)) {
        log_warning("Concatenated header value for %s is invalid, ignoring.", name);
        continue;
}
```

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,109 @@
+#include "strv.h"
+
+/* According to https://developers.cloudflare.com/rules/transform/request-header-modification/reference/header-format/
+HTTP header name can contain:
+- Alphanumeric characters: a-z, A-Z, and 0-9
+- The following special characters: - and _ */
```
Please insert ` *` at the beginning of each line.

yuwata commented on src/journal-remote/journal-header-util.c:
```diff
@@ -0,0 +1,109 @@
+#define HEADER_VALUE_LENGTH_MAX 8000
+
+/* According to https://developers.cloudflare.com/rules/transform/request-header-modification/reference/header-format/
+HTTP header value can contain:
+- Alphanumeric characters: a-z, A-Z, and 0-9
+- The following special characters: _ :;.,\/"'?!(){}[]@<>=-+*#$&`|~^%
```
Also here.

yuwata commented on man/journal-upload.conf.xml:
```diff
@@ -129,6 +129,28 @@
+        <listitem><para>Specifies an additional HTTP header to be added to each request to a URL.
+        Takes a pair of header name and value separated with a colon(<literal>:</literal>),
+        e.g. <literal>Name:Value</literal>.
+        Header name can contain alphanumeric values, "_" and "-" symbols additionally.
```
Please use `<literal>` tag for quotation: 
`<literal>_</literal> and <literal>-</literal>`

yuwata reviewed (changes requested):
Looks mostly OK. But one more round please.

yuwata reviewed (approved):
LGTM.
